numpy/core/records.py

```diff
@@ -423,6 +423,12 @@ def __new__(subtype, shape, dtype=None, buf=None, offset=0, strides=None,
                                       strides=strides, order=order)
         return self
 
+    def __array_finalize__(self, obj):
+        if type(obj) is not type(self):
+            # invoke __setattr__, which for void dtypes will do
+            # self.dtype = sb.dtype((record, self.dtype))
+            self.dtype = self.dtype
```

Comment:
@ahaldane - it should be:
```
class myarray(np.ndarray):
    def __array_finalize__(self, obj):
        print(self, obj)
        print(type(self), type(obj), type(self) is type(obj))
        if obj is not None:
            print(self.dtype, obj.dtype, self.dtype is obj.dtype)

# Now the three ways we can get to __array_finalize__
myarray((4,), buffer=np.arange(4.))
# [ 0.  1.  2.  3.] None
# <class '__main__.myarray'> <class 'NoneType'>
# -->  myarray([ 0.,  1.,  2.,  3.])
mya[:2]
# [ 0.  1.] [ 0.  1.  2.  3.]
# <class '__main__.myarray'> <class '__main__.myarray'>
# float64 float64 True
# --> myarray([ 0.,  1.])
np.arange(4.).view(myarray)
# [ 0.  1.  2.  3.] [ 0.  1.  2.  3.]
# <class '__main__.myarray'> <class 'numpy.ndarray'> False
# float64 float64 True
# --> Out[15]: myarray([ 0.,  1.,  2.,  3.])
```

I think this is just a problem with your `if` statement; it should be
```
if obj is not None and type(self) is not type(obj):
    self.dtype = obj.dtype
```